### ISSUE TITLE
BIPM review: math formulas are not copy/pastable

### ISSUE THREAD

**Issue by ronaldtse** [metanorma/bipm-si-brochure#118]
From BIPM:

> The observation with the largest number of repetitions is that numerical values, unit symbols and quantity symbols in the pdf cannot be selected and are not searchable. I imagine that this is related to the use of UnitsML for encoding in the XML file and the way this is represented in the pdf. Hopefully this can be easily solved.

This is similar to the requests here:
* https://tex.stackexchange.com/questions/352649/can-we-get-maths-in-pdfs-that-can-be-copy-pasted-and-screen-read
* https://tex.stackexchange.com/questions/233390/in-which-way-have-fake-spaces-made-it-to-actual-use/233397#233397

Which indicates that it is possible to provide different presentational content and copy/paste content in PDF.

The `accsupp` LaTeX package there is for accessibility support, and I found in its package this is a PDF feature called "Replacement Text":

https://github.com/ho-tex/accsupp/blob/14b25255f16281e02e50aa66a3632c39ec4c88f5/accsupp.dtx#L285-L301

```
% PDF specification says in ``10.8.3 Replacement Text''
% \begin{quote}
% Just as alternative descriptions can be provided for images and other
% items that do not translate naturally into text (\dots), replacement
% text can be specified for content that does translate into text but
% that is represented in a nonstandard way. These nonstandard representations
% might include, for example, glyphs for ligatures or custom characters, or
% inline graphics corresponding to letters in an illuminated manustript or to
% dropped capitals.
% \end{quote}
% However, the ActualText is ignored in Acrobat Reader
% (until version 9 at least), if the content does not contain glyphs.
% Option \xoption{space} adds such an invisible glyph, a space
% character. The font name can be configured by option \xoption{spacefont},
% the default font is \texttt{phvr8r}. The character slot is given
% by option \xoption{spacechar}, default is \texttt{32}, the usual
% position of the space character.
```

Could we utilize this PDF feature to insert the source AsciiMath / LaTeX math?

**Comment by Intelligent2013** [comment]
Apache FOP has a limitation (https://xmlgraphics.apache.org/fop/2.5/accessibility.html#limitations):
* SVG graphics (or images in general) are treated as a single figure. Text contained in SVGs is not accessible. It's only possible to work with fox:alt-text.

In the current version of xslt, the attribute `fox:alt-text` for mathml contains just word 'Math':
```xml
<fo:instream-foreign-object fox:alt-text="Math"><math ...
```
Internally in PDF it added as `/Alt`:
```
1910 0 obj
<<
  /S /Figure
  /P 1909 0 R
  /Alt (Math)
  /A << /BBox [18 46 5 7] /O /Layout >>
  /K [<<
  /Type /MCR
  /Pg 22 0 R
  /MCID 55
>>]
>>
endobj
```

But I can't find how this alternate text can be used/showed.
Read Out Loud feature in Adobe Acrobat Reader just read a text line until first image and then stop reading:
![image](https://user-images.githubusercontent.com/5559926/121937530-7262f680-cd53-11eb-8041-f5c3d724e31d.png)
Tooltip also doesn't showing.


Also, I've tried to add 'Actual Text` manually via Adobe Acrobat (full version, not Reader) as described here https://www.adobe.com/accessibility/products/acrobat/pdf-repair-add-alternative-text.html:
![image](https://user-images.githubusercontent.com/5559926/121951403-8e6e9400-cd63-11eb-9ef4-fb2487fe6032.png)

and after that the sentence `Example of actual text` I can see only in the text file generated by File->SaveAs->Text(Accessible).txt:
```txt
The BIPM has its headquarters near Paris, in the grounds (Example of actual text
) of the Pavillon de 
Breteuil (Parc de Saint-Cloud) placed at its disposal by the French Government; its upkeep is 
financed jointly by the Member States of the Metre Convention.
```
Again, I can't copy, find or view tooltip for the text `Example of actual text` in the Acrobat.

**Comment by ronaldtse** [comment]
@Intelligent2013 I think we have to do two things:

1. Set alt text (if we cannot set "actual text") for the SVG formulas to the AsciiMath/LaTeX math (for screen readers to read)
2. Create a PDF comment for each of the SVG formulas to AsciiMath/LaTeX math or MathML (for people to copy out formulas).

**Comment by Intelligent2013** [comment]
@ronaldtse 
>     1. Set alt text (if we cannot set "actual text") for the SVG formulas to the AsciiMath/LaTeX math (for screen readers to read)

Apache FOP doesn't support "actual text" (there isn't extension). We can set only "alt text" via `fox:alt-text` attribute.
@opoudjis can we add AsciiMath/LaTeX math in presentation xml? For example, in a tag like `asciimath`, or as comment immediaterly after/before mathml tag?

> 2. Create a PDF comment for each of the SVG formulas to AsciiMath/LaTeX math or MathML (for people to copy out formulas).

Apache FOP doesn't support comments/annotations. I think it could be done on additional post-processing step via PDFBox, but  I'm not sure what further issues we'll find in PDFBox....

**Comment by ronaldtse** [comment]
> @opoudjis can we add AsciiMath/LaTeX math in presentation xml? For example, in a tag like asciimath, or as comment immediaterly after/before mathml tag?

Yes, certainly.

**Comment by ronaldtse** [comment]
> Apache FOP doesn't support comments/annotations. I think it could be done on additional post-processing step via PDFBox, but I'm not sure what further issues we'll find in PDFBox....

We don't need this if we have the original text.

In this case if @opoudjis can provide the source math (AsciiMath or LaTeX math, depending on the original source), then we can embed this as alt-text, right? This means the reader could copy and paste this content, and it's sufficient.

**Comment by Intelligent2013** [comment]
> In this case if @opoudjis can provide the source math (AsciiMath or LaTeX math, depending on the original source), then we can embed this as alt-text, right? This means the reader could copy and paste this content, and it's sufficient.

Yes, we can embed AsciiMath/LaTeX  math as alt-text. But I don't know a simply way to copy the content of alt-text as I've described above, i.e. open PDF in Acrobat (full version), open Tags tab and find the math object (manually) in the tree. In the Acrobat there isn't context menu for math object that allow us to open the Object Properties quickly.

**Comment by ronaldtse** [comment]
I wonder if the correct approach is to set the Alt Text to source (AsciiMath/LaTeX math), then Actual Text as it is now, and if people really want to copy/paste, ask them to use the HTML copy.

**Comment by ronaldtse** [comment]
From @Intelligent2013 

> * in Acrobat Reader DC and Preview, I can't find a way to see alt text...

the same in my environment.

> * in Acrobat Pro DC, I need to use "Accessibility" pane's "Set Alternate Text" to see the text "Math" on the rendered math

the same in my environment, and I can't just choose any math object and set alternate text, but I  need to click Next button tens times to find the need object...

> * in Acrobat Pro DC, I need to use the "View > Show/Hide > Navigation Pane > Tags" pane to manually drill down

the same in my environment.

> * in Acrobat Pro DC, the "Read Out Loud" function reads the "Alternate Text", not the "Actual Text".

"Read Out Loud" function for math object works only when I choose 'Read This Page Only', or 'Read to End of Document`. But if I click on any text, then it starts read until the end of line (but sometimes stops in a middle, or before comma) or until math object....

**Comment by ronaldtse** [comment]
From @Intelligent2013 :

> One more idea - to add mathml file as attachment for each math, and put a link on math object to this attachment. User click on the formula, and Acrobat open attached file.
>
> In the attached example I've put just one mathml.xml for all formulas, just for experiment.
> 
> [si-brochure-en.presentation (2).pdf](https://github.com/metanorma/bipm-si-brochure/files/7004800/si-brochure-en.presentation.2.pdf)

**Comment by ronaldtse** [comment]
I've tried it on Preview and Acrobat Reader, in Preview it clicks but nothing happens, on Acrobat Reader it downloads mathml.xml. So maybe this is the best way ahead. The user must use Acrobat Reader in order to copy/paste the MathML.

**Comment by ronaldtse** [comment]
I looked into the W3C sample PDF which only contains content but no tags:
https://www.w3.org/WAI/WCAG20/Techniques/working-examples/PDF7/ocr-example.pdf

In this PDF, there are no tags -- i.e. no actual text, no alternate text. However, there is an image background and also "content" behind it. The "content" is what comes out from a copy and paste.

<img width="1072" alt="image" src="https://user-images.githubusercontent.com/11865/129853275-563b6d0e-ce2b-4b84-9ba0-1cab9a07137d.png">

When I look at the "Tagged" PDF:
https://www.w3.org/WAI/WCAG20/Techniques/working-examples/PDF7/ocr-example-tagged.pdf

I noticed that an image can be marked as an "artifact", which describes that the element is not involved in the textual content. I think this is a good description of the math -- it is not involved in textual flow. Instead, we should 

This made me think that we should:
* Make the Math an Artifact (so it doesn't involve text flow)
* Add the actual formula source text (e.g. asciimath) inside the content.

By using Acrobat "Edit Content" I was able to convert the Math "PathPathPath" into an Artifact, then add the text inside the `<P>` text node to include the math source.

<img width="1090" alt="Screenshot 2021-08-18 at 3 15 40 PM" src="https://user-images.githubusercontent.com/11865/129854643-c44e62cc-66c3-49db-8997-06da46999727.png">

Here you can see the overlapping text, top being the Artifact, bottom being the actual content. And I confirm that the text can be copied.

**Comment by ronaldtse** [comment]
And I was able to change a text to white color (in order to hide the "content text" supplied).
<img width="692" alt="Screenshot 2021-08-18 at 3 19 22 PM" src="https://user-images.githubusercontent.com/11865/129855131-38738878-85e8-4b7c-af6f-9608fddee4dc.png">

I couldn't get the order right when it comes to hiding Math text here:
<img width="624" alt="Screenshot 2021-08-18 at 3 21 38 PM" src="https://user-images.githubusercontent.com/11865/129855426-8f87f393-622a-4c7a-8e31-4c6f11effad8.png">

**Comment by ronaldtse** [comment]
Right now the content structure is as follows:
- flowing text (1)
- math object (2)
- flowing text (3)
<img width="1061" alt="Screenshot 2021-08-18 at 3 36 37 PM" src="https://user-images.githubusercontent.com/11865/129857457-bf0905ac-5e2c-4641-8782-4a234ad962d9.png">

In order to enable copyable text, we have to change to:
- flowing text (1)
- new element that is flowing text, but contains the source of the math object (e.g. AsciiMath) (2')
- flowing text (3)
- the math object as Artifact that is placed right above element 2', and hides element 2' (4)

**Comment by Intelligent2013** [comment]
In Apache FOP there is a featire 'Treating Content as Artifact', so we can mark the content as Artifact (http://xmlgraphics.apache.org/fop/2.4/accessibility.html):
![image](https://user-images.githubusercontent.com/5559926/129861087-d55f4e01-5b79-46a0-b24b-a68fc22dd01c.png)

**Comment by Intelligent2013** [comment]
So, the task:
1. determine position of math object in the tag tree
2. place new text object (transparency) with AsciiMath before math object.
But the main problem is - the width of text object can be wider than math object, or individual chars in text object may not be positioned in the same place as individual curve of math object, and when we'll find some char (Ctlf-F), then it may be highlighted in some another place than original glyph.

**Comment by ronaldtse** [comment]
@Intelligent2013 for step 1, if we can determine the height and width of math object, can we shrink the content of the "math text" into a box of a suitable size?

**Comment by Intelligent2013** [comment]
> 
> @Intelligent2013 for step 1, if we can determine the height and width of math object, can we shrink the content of the "math text" into a box of a suitable size?

We can't just resize the rectangle text, we should change the font-size. I.e. font-size of the string depends on the bounding box... I don't figure out how font-size can be calculated simply... 
I think it makes no sense. For example:
source adoc:
```adoc
1 "unitsml(K)" = ((1.380649 xx 10^(-23))/k) "unitsml(kg*m^2*s^(-2))"
```
rendered as:
![image](https://user-images.githubusercontent.com/5559926/129866159-0cacefe4-ae77-44cd-a513-89bea22961d6.png)

Let's assumed that we resize math text (blue colored text) and place it above math object (on the figure it placed below just for simplicity):

![image](https://user-images.githubusercontent.com/5559926/129867781-f5d15d59-418a-4829-9f7d-1e304eab1994.png)

If user will find:
- `K`, then middle of `(` will be highlighted
- `3`, then `8` will be highlighted
- `k`, then `0` will be highlighted

**Comment by ronaldtse** [comment]
@Intelligent2013 good point, but I'm not sure what else we can do... If the user highlights "10^-23", the user can't really copy the semantic portion anyway. I think the Attachment approach is also good, but it only works for Acrobat.

Maybe we should use Attachment, Alt Text, Actual Text and the Content techniques together.

**Comment by Intelligent2013** [comment]
ok, then next tasks are:
- [ ] investigate how to add transparency text behind/above math object in Apache FOP Intermediate Format (IF)
### else
-----------------------
- [ ] try to add transparency text with PDFBox
- [ ] solve issue 
![image](https://user-images.githubusercontent.com/5559926/127686197-1d4e330f-8083-42c7-971d-cb42737b907b.png)
(https://github.com/metanorma/mn2pdf/issues/68#issuecomment-890072855)
-----------------------
- [ ] mark the content as Artifact
- [ ] add math markup as Attachment

**Comment by ronaldtse** [comment]
Thank you @Intelligent2013 !

**Comment by ronaldtse** [comment]
In fact, @Intelligent2013, can you make a blog post about this since this is a set of excellent techniques for PDF accessibility? Thanks!

**Comment by Intelligent2013** [comment]
> 
> 
> In fact, @Intelligent2013, can you make a blog post about this since this is a set of excellent techniques for PDF accessibility? Thanks!

@ronaldtse ok, but for begin, let's  I'll realize these ideas in the practice...
